File: eco_restoration/data_science/canal_kriging_causal_20260817/java/OrdinaryKriging.java
import java.util.ArrayList;
import java.util.List;

public final class OrdinaryKriging {
    private OrdinaryKriging() {}

    private record Observation(double x, double y, double value) {}
    private record Result(double estimate, double variance, double weightSum) {}

    private static double semivariance(double h, double nugget, double sill, double range) {
        return h <= 0.0 ? 0.0 : nugget + sill * (1.0 - Math.exp(-h / range));
    }

    private static double distance(double x1, double y1, double x2, double y2) {
        double dx = x1 - x2;
        double dy = y1 - y2;
        return Math.sqrt(dx * dx + dy * dy);
    }

    private static double[] solve(double[][] matrix, double[] rhs) {
        int n = rhs.length;

        for (int column = 0; column < n; column++) {
            int pivot = column;
            for (int row = column + 1; row < n; row++) {
                if (Math.abs(matrix[row][column]) > Math.abs(matrix[pivot][column])) {
                    pivot = row;
                }
            }

            if (Math.abs(matrix[pivot][column]) <= 1.0e-12) {
                throw new IllegalArgumentException("singular kriging system");
            }

            double[] swapRow = matrix[pivot];
            matrix[pivot] = matrix[column];
            matrix[column] = swapRow;

            double swapValue = rhs[pivot];
            rhs[pivot] = rhs[column];
            rhs[column] = swapValue;

            double divisor = matrix[column][column];
            for (int col = column; col < n; col++) {
                matrix[column][col] /= divisor;
            }
            rhs[column] /= divisor;

            for (int row = 0; row < n; row++) {
                if (row == column) {
                    continue;
                }

                double factor = matrix[row][column];
                for (int col = column; col < n; col++) {
                    matrix[row][col] -= factor * matrix[column][col];
                }
                rhs[row] -= factor * rhs[column];
            }
        }

        return rhs;
    }

    private static Result interpolate(
        List<Observation> observations,
        double targetX,
        double targetY,
        double nugget,
        double sill,
        double range
    ) {
        if (observations.size() < 2 || nugget < 0.0 || sill < 0.0 || range <= 0.0) {
            throw new IllegalArgumentException("invalid kriging inputs");
        }

        int n = observations.size();
        double[][] matrix = new double[n + 1][n + 1];
        double[] rhs = new double[n + 1];

        for (int row = 0; row < n; row++) {
            Observation left = observations.get(row);
            for (int col = 0; col < n; col++) {
                Observation right = observations.get(col);
                matrix[row][col] = semivariance(
                    distance(left.x(), left.y(), right.x(), right.y()),
                    nugget, sill, range
                );
            }
            matrix[row][n] = 1.0;
            matrix[n][row] = 1.0;
            rhs[row] = semivariance(distance(left.x(), left.y(), targetX, targetY), nugget, sill, range);
        }

        rhs[n] = 1.0;
        double[] solution = solve(matrix, rhs);
        double estimate = 0.0;
        double variance = solution[n];
        double weightSum = 0.0;

        for (int index = 0; index < n; index++) {
            estimate += solution[index] * observations.get(index).value();
            variance += solution[index] * rhs[index];
            weightSum += solution[index];
        }

        return new Result(estimate, Math.max(0.0, variance), weightSum);
    }

    public static void main(String[] args) {
        if (args.length < 11 || (args.length - 5) % 3 != 0) {
            System.err.println(
                "usage: OrdinaryKriging <target_x_m> <target_y_m> <nugget> <sill> <range_m> " +
                    "<x_m> <y_m> <value> [<x_m> <y_m> <value> ...]"
            );
            System.exit(64);
        }

        try {
            double targetX = Double.parseDouble(args[0]);
            double targetY = Double.parseDouble(args[1]);
            double nugget = Double.parseDouble(args[2]);
            double sill = Double.parseDouble(args[3]);
            double range = Double.parseDouble(args[4]);

            List<Observation> observations = new ArrayList<>();
            for (int index = 5; index < args.length; index += 3) {
                observations.add(new Observation(
                    Double.parseDouble(args[index]),
                    Double.parseDouble(args[index + 1]),
                    Double.parseDouble(args[index + 2])
                ));
            }

            Result result = interpolate(observations, targetX, targetY, nugget, sill, range);
            System.out.printf("estimate=%.10f%n", result.estimate());
            System.out.printf("kriging_variance=%.10f%n", result.variance());
            System.out.printf("weight_sum=%.10f%n", result.weightSum());
            System.out.println(
                "unbiasedness_constraint=" +
                    (Math.abs(result.weightSum() - 1.0) <= 1.0e-8 ? "SATISFIED" : "FAILED")
            );
        } catch (IllegalArgumentException exception) {
            System.err.println("error: " + exception.getMessage());
            System.exit(65);
        }
    }
}

File: eco_restoration/data_science/canal_kriging_causal_20260817/cpp/ordinary_kriging.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <string>
#include <vector>

struct Observation {
    double x_m;
    double y_m;
    double value;
};

static double distance(const Observation& a, const Observation& b) {
    const double dx = a.x_m - b.x_m;
    const double dy = a.y_m - b.y_m;
    return std::sqrt(dx * dx + dy * dy);
}

static double distance_to_target(const Observation& observation, double x_m, double y_m) {
    const double dx = observation.x_m - x_m;
    const double dy = observation.y_m - y_m;
    return std::sqrt(dx * dx + dy * dy);
}

static double semivariance(double h, double nugget, double sill, double range_m) {
    if (h <= 0.0) {
        return 0.0;
    }
    return nugget + sill * (1.0 - std::exp(-h / range_m));
}

static std::vector<double> solve(std::vector<std::vector<double>> matrix, std::vector<double> rhs) {
    const int n = static_cast<int>(rhs.size());

    for (int column = 0; column < n; ++column) {
        int pivot = column;
        for (int row = column + 1; row < n; ++row) {
            if (std::abs(matrix[row][column]) > std::abs(matrix[pivot][column])) {
                pivot = row;
            }
        }

        if (std::abs(matrix[pivot][column]) <= 1.0e-12) {
            throw std::runtime_error("singular kriging system; revise points or variogram parameters");
        }

        std::swap(matrix[pivot], matrix[column]);
        std::swap(rhs[pivot], rhs[column]);

        const double divisor = matrix[column][column];
        for (int col = column; col < n; ++col) {
            matrix[column][col] /= divisor;
        }
        rhs[column] /= divisor;

        for (int row = 0; row < n; ++row) {
            if (row == column) {
                continue;
            }

            const double factor = matrix[row][column];
            for (int col = column; col < n; ++col) {
                matrix[row][col] -= factor * matrix[column][col];
            }
            rhs[row] -= factor * rhs[column];
        }
    }

    return rhs;
}

static void ordinary_kriging(
    const std::vector<Observation>& observations,
    double target_x_m,
    double target_y_m,
    double nugget,
    double sill,
    double range_m,
    double& estimate,
    double& variance,
    double& weight_sum
) {
    if (observations.size() < 2 || nugget < 0.0 || sill < 0.0 || range_m <= 0.0) {
        throw std::invalid_argument("need at least two observations, nugget/sill >= 0, and positive range");
    }

    const int n = static_cast<int>(observations.size());
    std::vector<std::vector<double>> matrix(n + 1, std::vector<double>(n + 1, 0.0));
    std::vector<double> rhs(n + 1, 0.0);

    for (int row = 0; row < n; ++row) {
        for (int col = 0; col < n; ++col) {
            matrix[row][col] = semivariance(distance(observations[row], observations[col]), nugget, sill, range_m);
        }
        matrix[row][n] = 1.0;
        matrix[n][row] = 1.0;
        rhs[row] = semivariance(distance_to_target(observations[row], target_x_m, target_y_m), nugget, sill, range_m);
    }

    matrix[n][n] = 0.0;
    rhs[n] = 1.0;

    const std::vector<double> solution = solve(matrix, rhs);
    estimate = 0.0;
    variance = solution[n];
    weight_sum = 0.0;

    for (int index = 0; index < n; ++index) {
        estimate += solution[index] * observations[index].value;
        variance += solution[index] * rhs[index];
        weight_sum += solution[index];
    }

    if (variance < 0.0 && variance > -1.0e-9) {
        variance = 0.0;
    }
}

static double leave_one_out_rmse(
    const std::vector<Observation>& observations,
    double nugget,
    double sill,
    double range_m
) {
    if (observations.size() < 3) {
        return std::numeric_limits<double>::quiet_NaN();
    }

    double squared_error = 0.0;
    for (std::size_t held_out = 0; held_out < observations.size(); ++held_out) {
        std::vector<Observation> retained;
        retained.reserve(observations.size() - 1);

        for (std::size_t index = 0; index < observations.size(); ++index) {
            if (index != held_out) {
                retained.push_back(observations[index]);
            }
        }

        double estimate = 0.0;
        double variance = 0.0;
        double weight_sum = 0.0;
        ordinary_kriging(
            retained,
            observations[held_out].x_m,
            observations[held_out].y_m,
            nugget,
            sill,
            range_m,
            estimate,
            variance,
            weight_sum
        );

        const double residual = observations[held_out].value - estimate;
        squared_error += residual * residual;
    }

    return std::sqrt(squared_error / observations.size());
}

int main(int argc, char** argv) {
    if (argc < 11 || ((argc - 6) % 3 != 0)) {
        std::cerr
            << "usage: " << argv[0]
            << " <target_x_m> <target_y_m> <nugget> <sill> <range_m>"
            << " <x_m> <y_m> <value> [<x_m> <y_m> <value> ...]\n";
        return 64;
    }

    try {
        const double target_x_m = std::stod(argv[1]);
        const double target_y_m = std::stod(argv[2]);
        const double nugget = std::stod(argv[3]);
        const double sill = std::stod(argv[4]);
        const double range_m = std::stod(argv[5]);

        std::vector<Observation> observations;
        for (int index = 6; index < argc; index += 3) {
            observations.push_back({
                std::stod(argv[index]),
                std::stod(argv[index + 1]),
                std::stod(argv[index + 2])
            });
        }

        double estimate = 0.0;
        double variance = 0.0;
        double weight_sum = 0.0;
        ordinary_kriging(
            observations, target_x_m, target_y_m, nugget, sill, range_m,
            estimate, variance, weight_sum
        );

        const double rmse = leave_one_out_rmse(observations, nugget, sill, range_m);
        const double knowledge_factor = observations.size() >= 6 ? 0.75 : 0.45;
        const double harm_risk = observations.size() >= 6 && std::isfinite(rmse) ? 0.30 : 0.65;
        const double eco_impact_value = knowledge_factor * (1.0 - harm_risk);

        std::cout << std::fixed << std::setprecision(10);
        std::cout << "estimate=" << estimate << '\n';
        std::cout << "kriging_variance=" << variance << '\n';
        std::cout << "weight_sum=" << weight_sum << '\n';
        std::cout << "leave_one_out_rmse=";
        if (std::isfinite(rmse)) {
            std::cout << rmse << '\n';
        } else {
            std::cout << "UNAVAILABLE\n";
        }
        std::cout << "knowledge_factor=" << knowledge_factor << '\n';
        std::cout << "eco_impact_value=" << eco_impact_value << '\n';
        std::cout << "harm_risk=" << harm_risk << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 65;
    }
}
